### PR TITLE
Update search method in menus

### PR DESCRIPTION
**Pull request by aidenlx**
1. Utilize obsidian's built-in fuzzy search method, used in quick switcher
2. With [cm-chs-patch v1.3.1+](https://github.com/aidenlx/cm-chs-patch) installed, the search method will use Chinese parse API to convert Chinese into Pinyin, which can be searched by key strokes

### CODE REVIEW

File: src/menus.ts [hunk 1]
```diff
@@ -103,25 +103,25 @@ export class PopupMenu extends Menu {
...
-
-    find(pattern: RegExp) {
-        let pos = Math.min(0, this.selected);
+    searchFor(match: string): boolean {
+        const query = prepareQuery(match);
+        let pos = Math.min(0, this.selected), 
+            results = [] as (SearchResultContainer & { pos: number })[];
         for (let i=this.items.length; i; ++pos, i--) {
             if (this.items[pos]?.disabled) continue;
-            if (this.items[pos]?.dom.textContent.match(pattern)) {
```
Comment (aidenlx) [inline src/menus.ts:119]: Is this intentional to get all texts, including extensions located in `.nav-file-tag`?

**Comment by pjeby** [comment]
Before even reviewing the code I need to understand what this is *for*.  What is the current search not doing that you do want, or is doing that you don't want?  I am not sufficiently clear on how the Obsidian fuzzy search algorithm works to know if it's what I would like to use or would be an improvement on the current algorithm, which I have written to be easy to reason about, and to minimize the number of keystrokes when using generally-stable lists (like context menus).  Obsidian's fuzzy search is useful when you are looking at the screen, but I find its behavior hard to predict in the presence of minor changes to a list's contents.

Regarding the file names, being able to search file extensions is indeed intentional.

Anyway, for future reference, when submitting pull requests to projects it is a good idea to explain up front what problem is being solved or what you are trying to achieve, so the maintainer doesn't have to try to guess.  In this case I can guess is that you want to make searching for chinese characters better, but it seems like using the fuzzy search may be counterproductive for that, whereas it might be possible to improve the current algorithm's regex generation for pinyin.

To do so, however, I'd need to know how you actually expect to search.  Do you intend to type the full character names?  With spaces between?  Or do you intend to use just parts of the character names?

**Comment by aidenlx** [comment]
One of the benefits of using a fuzzy search algorithm is that it can specify weighting manually, meaning that it can match filename first, then trying to match extensions, but digging into the algorithm will definitely take more time, and since reverse-engineering the code for quick switcher is already tiresome, I'll revert that change and only keep Chinese support for now. 

 > Do you intend to type the full character names? With spaces between? Or do you intend to use just parts of the character names?

I'm adding space between pinyins because I'm trying to mimic English words so that it may work better for the fuzzy search algorithm, and since I'm reverting it, it may not be necessary. 

And regarding using keystrokes to search for Chinese, the usual ways are searching by the first letter of pinyin, for example, `今(jin)天(tian)` ->`jt`, or by entering full pinyin (`今(jin)天(tian)`->`jint`). But it may not be so simple when polyphone exists, meaning that one character in Chinese can have different pinyin in various contexts, like `长` in `长(zhang)大` and `长(chang)期`. The library I'm using is relatively lightweight, so it won't try to predict the exact pinyin for the character, instead, it will give all the possible pinyin. In `长大(da)`'s case, it will produce [["zhang", "chang"],["da"]], and the current solution is flatting the list and join the members, like `zhangchangda`. 

But here is the problem: with default sorting, the Chinese names are always before English ones, so with the old matching method which uses the sequence of `item` to get the first match, the result will always be Chinese first. In my mind, the solutions to may be mapping item list to name list and resorting the mapped one with pinyins, and it will take some time for me to work on the details. To be honest, fuzzy search algorithm may be the ultimate solution across languages, if I have time to understand what it does...

btw, do you need the new `yarn.lock` file committed? it seems that it has been largely changed after adding a new module.

**Comment by pjeby** [comment]
Pull requests are not for "here is some code that may or may not do something vaguely related to what I want, which I don't actually know precisely what I want yet.  But please put this code into your project".

When you submit a pull request to an open source project, you are requesting a *pull*, implying that the code you have is *ready* to be pulled and thus incorporated into the project's code.  This implies that you have:

* Thought out what it is supposed to do
* Considered how it will impact other users of the project
* Considered the support and maintenance impact for the core contributors
* Written the code according to the project's style

And above all, actually *finished the thing you're trying to do*.  At a bare minimum, if the thing is not finished, you should give a clear indication *in the subject line* that it is a work in progress, and most open source projects do not accept WIP PRs that aren't targeted at development branches or submitted by trusted contributors.

What you have basically done is submit a PR where you have no specific intention beyond a vague idea that Obsidian's fuzzy search will somehow be beneficial, with none of the research or testing or evidence as to why this should be so, and are implicitly requesting this code be added to the project.  In effect, you are telling me that you consider this state of things good enough for me to add this code to the project.

I was trying to give you the benefit of the doubt by assuming that you had in fact thought this through, and simply hadn't communicated things that must be obvious to you and assumed I would know them.  But now I see that it's rather that you haven't thought this through at all, and are instead simply doing whatever seems like a good idea at the time.

This tells me that I cannot trust your judgment.  You keep saying you want to collaborate, but your behavior is not that of someone who is trying to earn the trust necessary for collaboration.  If everything you submit is suspect as to whether you've given any basic thought to what you're proposing, or basic courtesies like communicating intended results or what impacts you considered, I cannot afford the time to keep entertaining your submissions and will ultimately begin blocking your communications.

I remain open to incorporating some sort of search improvement for CJK characters, but not in the form of incomplete PRs for undefined requirements.

Even for bug reports or feature requests, it is expected that you at least be able to communicate clearly what behavior you want the improvement to have...  implying that you have already done the work to figure out what it is that you want.  Posting issues or PRs where you aren't even close to having an idea of what it is you want is just wasting people's time.  Doing it in *code* is even more wasteful, since code inherently requires more time to process than text.

Please do not submit any more PRs to this project until and unless they meet the requirements above.  If you believe at some point you have come up with a strategy for CJK or pinyin search -- one that includes the expected impact on other users -- then feel free to open an issue describing what you want to do and how it should work, preferably with examples.

It is often a good idea, by the way, to submit an issue for something before submitting a PR, especially when contributing to a project that has not previously accepted PRs from you.  In this way, you can learn whether it is even worth doing the work to write the PR, in case your desired change is one the project team considers out of scope, counterproductive, something they would not wish to maintain, or something they would have specific requirements for that you would otherwise not know about.

Having such a discussion first is a sign of respect, that you understand that you are neither an expert nor an authority within that project's community, whereas jumping straight to a PR that isn't tied to an existing issue is an implicit statement that you understand the project's plans, processes, and requirements as well as an existing team member -- which is very rarely the case.

In essence, when you submit a PR without an issue, you are staking your reputation on the idea that in that particular area, you are as good as the project team.  For things that are minor bug fixes or unequivocal improvements this may well be the case.  But if you guess wrongly, then you are showing that not only do you not understand the project as well as you think, but that you also *think* you're better at it than you actually are.  And that damages others' opinion not only of your ability, but also of your wisdom, forethought, and judgment.

I am closing this PR.  Please do not open any more PRs to any of my projects without first submitting an issue that describes clearly how you would like that project's behavior to change, and what impacts you expect that to have on other users...  and then wait for an explicit approval before submitting a PR.

In other words, I am no longer giving you the benefit of the doubt, because you have shown at this point that I cannot trust your judgment.

**Comment by aidenlx** [comment]
Lesson learnt，thank you so much for all the time you've spent to help me finally understand what an Pull Request meant for an project and the principles of collaboration in open source community, I'll follow your advice on this for every PR I opened in the future, you are such a patient and wonderful tutor for me, really sorry for having you tolerating all my terrible inappropriateness and ignorance during our conversations.

**Comment by pjeby** [comment]
Ignorance isn't a problem: I'm happy to answer questions and even offer advice.  It's the lack of *asking* (or admitting to ignorance) that's been much of the problem.  It would've wasted a lot less time if e.g. you said in an issue, "I need to do this kind of search and the existing search doesn't work well in this way...  do you think that using Obsidian's fuzzy search would help?"  Then you could have saved time writing the code and I would've saved time reading it, because we could've discussed it first.

I am actually still curious about solving the problem itself -- for example, I'm wondering if you find Obsidian's "quick open" dialog to be useful in searching for notes using the approach you describe, and if not, how you currently find such notes.  If the quick open dialog is useful, on the other hand, then perhaps it would make sense to allow opening quick explorer on a note from the quick open.  And so on.  There are many things that could be discussed to find a good alternative, though of course there is no guarantee that any of them would work out.

Also, not every project maintainer is as particular about PRs as I am, but in general the larger the project or team, the greater the odds that they will have very specific expectations for PRs and prefer to see issues first from new contributors.  And most will expect works in progress to be labeled as such.
